File: 113-path-sum-ii/113-path-sum-ii.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def pathSum(self, root: Optional[TreeNode], targetSum: int) -> List[List[int]]:
        output = []
        def dfs(root, path, currSum):
            if root:
                path.append(root.val)
                currSum += root.val
                if not root.right and not root.left and currSum == targetSum:
                    output.append(path[:])
                if root.left:
                    dfs(root.left, path, currSum)
                if root.right:
                    dfs(root.right, path, currSum)
                path.pop()
                currSum -= root.val
        dfs(root, [], 0)
        return output

File: 113-path-sum-ii/113-path-sum-ii.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func pathSum(root *TreeNode, targetSum int) [][]int {
    result:=make([][]int,0)
    tempArr := make([]int,0)
    dfs(root,targetSum,&result,tempArr)
    return result
}

func dfs(root *TreeNode, sum int, result *[][]int, tempArr []int){
    if root == nil{
        return
    }
    if root.Left == nil && root.Right == nil && root.Val == sum{
        tempArr = append(tempArr, root.Val)
        *result = append(*result, append([]int(nil),tempArr...))
    }
    dfs(root.Left, sum -root.Val, result, append(tempArr, root.Val))
    dfs(root.Right, sum -root.Val, result, append(tempArr, root.Val))
}